model/jpa/src/main/java/org/keycloak/models/jpa/JpaIDPProvider.java

```diff
@@ -0,0 +1,333 @@
+        IdentityProviderEntity entity = this.getEntityById(identityProvider.getInternalId(), true);
+        entity.setAlias(identityProvider.getAlias());
+        entity.setDisplayName(identityProvider.getDisplayName());
+        entity.setEnabled(identityProvider.isEnabled());
+        entity.setTrustEmail(identityProvider.isTrustEmail());
+        entity.setAuthenticateByDefault(identityProvider.isAuthenticateByDefault());
+        entity.setFirstBrokerLoginFlowId(identityProvider.getFirstBrokerLoginFlowId());
+        entity.setPostBrokerLoginFlowId(identityProvider.getPostBrokerLoginFlowId());
+        entity.setAddReadTokenRoleOnCreate(identityProvider.isAddReadTokenRoleOnCreate());
+        entity.setStoreToken(identityProvider.isStoreToken());
+        entity.setConfig(identityProvider.getConfig());
+        entity.setLinkOnly(identityProvider.isLinkOnly());
+
+        // send identity provider updated event.
+        RealmModel realm = this.getRealm();
+        session.getKeycloakSessionFactory().publish(new RealmModel.IdentityProviderUpdatedEvent() {
+
+            @Override
+            public RealmModel getRealm() {
+                return realm;
+            }
+
+            @Override
+            public IdentityProviderModel getUpdatedIdentityProvider() {
+                return identityProvider;
+            }
+
+            @Override
+            public KeycloakSession getKeycloakSession() {
+                return session;
+            }
+        });
+    }
+
+    @Override
+    public boolean remove(String alias) {
+        // find provider by alias in the DB and remove it.
+        IdentityProviderEntity entity = this.getEntityByAlias(alias);
+
+        if (entity != null) {
+            em.remove(entity);
+
+            // send identity provider removed event.
+            RealmModel realm = this.getRealm();
+            IdentityProviderModel model = toModel(entity);
+            session.getKeycloakSessionFactory().publish(new RealmModel.IdentityProviderRemovedEvent() {
+
+                @Override
+                public RealmModel getRealm() {
+                    return realm;
+                }
+
+                @Override
+                public IdentityProviderModel getRemovedIdentityProvider() {
+                    return model;
+                }
+
+                @Override
+                public KeycloakSession getKeycloakSession() {
+                    return session;
+                }
+            });
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void removeAll() {
+        CriteriaBuilder builder = em.getCriteriaBuilder();
+        CriteriaDelete<IdentityProviderEntity> delete = builder.createCriteriaDelete(IdentityProviderEntity.class);
+        Root<IdentityProviderEntity> idp = delete.from(IdentityProviderEntity.class);
+        delete.where(builder.equal(idp.get("realmId"), this.getRealm().getId()));
+        this.em.createQuery(delete).executeUpdate();
```

Comment:
We are not publishing any `IdentityProviderRemovedEvent` in case of bulk removal. I'm not sure about all consequences, so I wanted to just highlight it.

Reply:
But when we implement `InfinispanIDPProvider` and we want to invalidate all cache entries we need to do it one-by-one anyway: `getAllStream().map(IdentityProviderModel::getAlias).forEach(this::remove);` so we never reach this method at all (in case `InfinispanIDPProvider` would be enabled).

Reply:
Good point.. I wanted to avoid iterating through every provider and calling remove on it. Question would be: this is called atm only during realm removal - who would be listening to these events during realm removal?

Reply:
@vramik Fire that event for each broker when removing a realm won't scale. It is the same approach we have now for groups, we just fire the realm removed event. Listening this event is more than enough to understand that the realm and any child is being removed.

Reply:
@vramik It depends on how you are caching entries. If you are adding entries that implement `InRealm` they will be automatically removed from the cache when running the realm invalidations.